utils_nlp/models/xlnet/sequence_classification.py

```diff
@@ -0,0 +1,363 @@
+        input_mask_tensor = torch.tensor(input_mask, dtype=torch.long)
+        labels_tensor = torch.tensor(labels, dtype=torch.long)
+
+        val_token_ids_tensor = torch.tensor(val_token_ids, dtype=torch.long)
+        val_input_mask_tensor = torch.tensor(val_input_mask, dtype=torch.long)
+        val_labels_tensor = torch.tensor(val_labels, dtype=torch.long)
+
+        if token_type_ids:
+            token_type_ids_tensor = torch.tensor(token_type_ids, dtype=torch.long)
+            val_token_type_ids_tensor = torch.tensor(val_token_type_ids, dtype=torch.long)
+
+            train_dataset = TensorDataset(
+                token_ids_tensor, input_mask_tensor, token_type_ids_tensor, labels_tensor
+            )
+
+            val_dataset = TensorDataset(
+                val_token_ids_tensor,
+                val_input_mask_tensor,
+                val_token_type_ids_tensor,
+                val_labels_tensor,
+            )
+
+        else:
+
+            train_dataset = TensorDataset(token_ids_tensor, input_mask_tensor, labels_tensor)
+
+            val_dataset = TensorDataset(
+                val_token_ids_tensor, val_input_mask_tensor, val_labels_tensor
+            )
+
+        # define optimizer and model parameters
+        param_optimizer = list(self.model.named_parameters())
+        no_decay = ["bias", "LayerNorm.weight"]
+        optimizer_grouped_parameters = [
+            {
+                "params": [p for n, p in param_optimizer if not any(nd in n for nd in no_decay)],
+                "weight_decay": self.weight_decay,
+            },
+            {
+                "params": [p for n, p in param_optimizer if any(nd in n for nd in no_decay)],
+                "weight_decay": 0.0,
+            },
+        ]
+
+        train_sampler = RandomSampler(train_dataset)
+        val_sampler = RandomSampler(val_dataset)
+
+        train_dataloader = DataLoader(
+            train_dataset, sampler=train_sampler, batch_size=self.batch_size
+        )
+
+        val_dataloader = DataLoader(val_dataset, sampler=val_sampler, batch_size=self.batch_size)
+
+        num_batches = len(train_dataloader)
+        num_train_optimization_steps = num_batches * self.num_epochs
+
+        optimizer = AdamW(optimizer_grouped_parameters, lr=self.lr, eps=self.adam_eps)
+        scheduler = WarmupLinearSchedule(
+            optimizer, warmup_steps=self.warmup_steps, t_total=num_train_optimization_steps
+        )
+
+        global_step = 0
+        self.model.train()
+        optimizer.zero_grad()
+        for epoch in range(self.num_epochs):
+
+            train_sampler = RandomSampler(train_dataset)
+
+            train_dataloader = DataLoader(
+                train_dataset, sampler=train_sampler, batch_size=self.batch_size
+            )
+
+            tr_loss = 0.0
+            logging_loss = 0.0
+            val_loss = 0.0
+
+            for i, batch in enumerate(tqdm(train_dataloader, desc="Iteration")):
```

Comment:
this code looks pretty similar to what we have in BERT: https://github.com/microsoft/nlp/blob/master/utils_nlp/models/bert/sequence_classification.py#L165

do you know which code is similar?

Once this is identified, we could refactor some functions to be used by both classes. And then we can even start to separate the content and so single responsibility as we discussed in the past with @eedeleon. In general, in other ML pieces of code we will have a signature like this:
```
def fit(self, data_loader_train, optimizer, data_loader_val=None, scheduler=None, n_epochs=10, ...)
```

Reply:
we're working on a new set of classes to unify the usage for all transformers in pytorch-transformers. Might be better to merge those first once they are done.